Add did_you_mean sources, gemspec, and load

bjfish opened the pull request:
For example,
```
$ jt p '"a".upcas'
$ /Users/brandonfish/Documents/jruby/bin/truffleruby -Xcore.load_path=/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby -Xgraal.warn_unless=false -e 'p begin "a".upcas end'
-e:1:in `method_missing': undefined method `upcas' for "a":String
Did you mean?  upcase
               upcase! (NoMethodError)
	from -e:1:in `<main>'
```
Is there a place I need to track license stuff for adding this gem? (MIT License)

After adding this, an interesting error occurs for `jt test fast`:
```
1)
Module#const_missing raises NameError and includes the name of the value that wasn't found ERROR
RubyTruffleError: internal implementation error - 
AssertionError <no message> org.truffleruby.core.symbol.SymbolTable.getSymbol(SymbolTable.java:74)
/Users/brandonfish/Documents/jruby/lib/mri/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:19:in `constants'
/Users/brandonfish/Documents/jruby/lib/mri/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:19:in `block in class_names'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/enumerable.rb:147:in `block in flat_map'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/enumerable.rb:146:in `each'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/enumerable.rb:146:in `flat_map'
/Users/brandonfish/Documents/jruby/lib/mri/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:18:in `class_names'
/Users/brandonfish/Documents/jruby/lib/mri/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:14:in `candidates'
/Users/brandonfish/Documents/jruby/lib/mri/did_you_mean/spell_checkable.rb:9:in `corrections'
/Users/brandonfish/Documents/jruby/lib/mri/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:26:in `corrections'
/Users/brandonfish/Documents/jruby/lib/mri/did_you_mean/core_ext/name_error.rb:20:in `corrections'
/Users/brandonfish/Documents/jruby/lib/mri/did_you_mean/core_ext/name_error.rb:13:in `to_s'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/exception.rb:44:in `message'
/Users/brandonfish/Documents/jruby/spec/ruby/core/module/const_missing_spec.rb:14:in `block (2 levels) in <main>'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/enumerable.rb:469:in `block in all?'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/enumerable.rb:469:in `each'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/enumerable.rb:469:in `all?'
/Users/brandonfish/Documents/jruby/spec/ruby/core/module/const_missing_spec.rb:4:in `<main>'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/kernel.rb:326:in `load'
/Users/brandonfish/Documents/jruby/truffleruby/src/main/ruby/core/kernel.rb:326:in `load'
```
The bad key in question looks like this:
```
CS_CONSTλ
0 = 'C' 67
1 = 'S' 83
2 = '_' 95
3 = 'C' 67
4 = 'O' 79
5 = 'N' 78
6 = 'S' 83
7 = 'T' 84
8 = 'λ' 955
```

It seems like an assertion is needed earlier to assert a bad key is not created/inserted into the symbolMap.

eregon commented on truffleruby/src/main/ruby/post-boot/shims.rb:
```diff
@@ -13,4 +13,12 @@ def self.default_dir
+    require 'did_you_mean'
+  rescue Gem::LoadError, LoadError => e
+  end # if defined?(DidYouMean) # TODO BJF 29-Jan-2017 The definition of this constant is used to
+                                # control the ruby option to disable did_you_mean
```
This should not go into `shims.rb` but in `gems.rb`.

bjfish replied:
@eregon I tried gems.rb first but this code is dependendent on the shim above it in shims.rb

eregon replied:
Argh. How about making a single consolidated `post-boot/post-boot.rb` then that integrate both files?

bjfish replied:
@eregon I've integrated into `post-boot` per your suggestion. I'm hoping this will resolve the build failure as well.

bjfish commented:
@chrisseaton I used version `1.0.0` because that is what is used in ruby `2.3.x`: https://github.com/ruby/ruby/blob/v2_3_3/gems/bundled_gems

MRI doesn't have the sources in their repository so I used the sources from a local `gem install did_you_mean -v 1.0.0`. Although, MRI would also download these sources locally as part of its `make dist` command.

eregon commented:
Re the assertion, it should create a Symbol with UTF-8 encoding. Fixed in afda54a1c66ac09b7dbbdd882eaa9c778654f390, please rebase.

bjfish commented:
@eregon Thanks, I've rebased

bjfish commented:
@chrisseaton I added the command line option, correctly, I hope

eregon commented:
Should we put the did_you_mean files next to other gems, in `lib/ruby/gems/shared` ? It seems bundled gems can be uninstalled, updated, etc like normal gems.

chrisseaton commented:
I'm not exactly sure what is best for these bundled gems. I think we want to vendor them, as you've done here, but I think we probably also want to separate them a bit from the normal stdlib. At the moment if I tried to update stdlib from MRI's repository it would look like MRI used to have these files and doesn't now. Could we have the full repo of the gem in `lib/gems/did_you_mean`. That will include the licence file. So just delete `.git`, `.ruby-version`, `.travis.yml` etc. Then you'll need to add each gem we include like that onto the load path.

Benoit does it need to be `lib/ruby/gems/shared`? Could it be just `lib/gems`?

eregon commented:
Alternatively we can put them in another "gem home" by using the GEM_PATH env var or similar mechanism, but it seems more fragile.

chrisseaton commented:
`lib/ruby/gems/shared` just strikes me as a strangely long path. What's the point of the `ruby` and `shared` parts for example?

eregon commented:
Adding them ourselves on the $LOAD_PATH might simplify things, but it's not the same as how RubyGems does it so there might be some issues.

eregon commented:
> lib/ruby/gems/shared just strikes me as a strangely long path. What's the point of the ruby and shared parts for example?

We kept it from JRuby, we could change it of course.

bjfish commented:
I've moved this bundled gem to `lib/gems/gems/did_you_mean` and the default gem specs to `lib/gems/specifications/default` and removed it from `stdlib`

bjfish commented:
This is still a work in progress. Moving the library out of stdlib has surfaced an issue with requiring the lib due to our default_dir shim.  Before shimming (when rubygems initializes)  the default dir points to `/Users/brandonfish/Documents/jruby/lib/ruby/gems/2.2.0/` so the default specs aren't initializing correctly. I'll try to figure this out.

chrisseaton commented:
I still don't understand why it's `lib/gems/gems`. Why `gems` twice?

bjfish commented:
@chrisseaton The `Gem.default_dir` looks like it has the usually structure of a `GEM_HOME` folder.
It has the sub-folders:
```
build_info	
cache		
doc		
extensions	
gems		
specifications
```

eregon commented on .gitignore:
```diff
@@ -52,6 +52,13 @@ lib/*.jar
+!lib/gems/gems/did_you_mean-1.0.0
+!lib/gems/specifications
+lib/gems/specifications/*
+!lib/gems/specifications/default
```
This seems rather complicated, can we just:
```
lib/gems
!lib/gems/gems/did_you_mean-1.0.0
!lib/gems/specifications/default
```

bjfish replied:
At least in my local version of git `2.7.1` 
If you do `lib/gems`, it never evens looks inside this folder to find the unexclude:
`!lib/gems/gems/did_you_mean-1.0.0`

so if you do `lib/gems/**` your not excluding the `lib/gem` folder just all the files/folders inside of it so it can see/find the unexclude on the next line.

Please correct me if I am wrong though because I was hoping to do something that simple

eregon replied:
Indeed, man gitignore says:
```
An optional prefix "!" which negates the pattern; any matching file excluded by a previous pattern will become included again. It
           is not possible to re-include a file if a parent directory of that file is excluded. Git doesn't list excluded directories for
           performance reasons, so any patterns on contained files have no effect, no matter where they are defined.
```

eregon commented:
@bjfish How did you add those files for did_you_mean? I think the cleanest would be to install it with RubyGems, like any other gem and check the files in as currently.

bjfish commented:
Here are the updates I've made since the last comments:

@eregon Before I was grabbing a copy of the Gem sources from my 2.3.1 GEM_HOME.
I took your advice and installed using the following:
`./bin/ruby -rbundler-workarounds -S gem install did_you_mean -v 1.0.0 --no-doc`
Then removed .gitignore, .travis.yml, and did-you-mean.gemspec

@chrisseaton I updated the directory structure to use: `lib/ruby/gems/2.3.0` which is the standard layout for this so I could remove the shim for `Gem.default_dir` and just use the normal `Gem.default_dir`.  I like that we can remove the shim for this and I believe it will save us problems in the future by following this layout. I could change this back to `lib/gems` if you'd like.

I learned that once `did_you_mean` was moved out of `stdlib` it is no longer considered a "default" gem and is now a "bundle" gem. So, I had to remove the `did_you_mean` gemspec from the `specifications/default` to the `specifications` folder otherwise it wouldn't be added to the path.

For reference, here are MRI's default vs bundle gems:
```
default gems:      
                              bigdecimal 1.2.8
                              io-console 0.4.5
                              json 1.8.3
                              psych 2.1.0
                              rdoc 4.2.1
bundle gems:      
                              did_you_mean 1.0.0
                              minitest 5.8.5
                              net-telnet 0.1.1
                              power_assert 0.2.6
                              rake 10.4.2
                              test-unit 3.1.5
```

I added a `did-you-mean` alias of `did_you_mean` in the command line parser, e.g. `--disable-did-you-mean`. I saw that was done for `frozen_string_literal`.

eregon commented:
Looks good. IMHO you could keep the .gitignore, .travis.yml, and did-you-mean.gemspec if they are shipped with the gem (but it's not a big deal).

Re structure, it's quite nice to remove the shim (could we configure it in some way or is that hardcoded in RubyGems?). The `2.3.0` component is not super useful, as it mean we'll rename the whole dir on every minor version. But removing the shim wins over that.
